src/utils/collectibles.utils.ts

```diff
@@ -0,0 +1,163 @@
+import { isNonEmptyArray } from '@apollo/client/utilities';
+import { Observable, catchError, map, of, withLatestFrom } from 'rxjs';
+
+import { ADULT_CONTENT_TAGS } from '../apis/objkt/adult-tags';
+import { ADULT_ATTRIBUTE_NAME } from '../apis/objkt/constants';
+import {
+  fetchGalleryAttributeCount$,
+  fetchFA2AttributeCount$,
+  fetchAllCollectiblesDetails$
+} from '../apis/objkt/index';
+import { CollectibleAttributes, CollectibleTag } from '../apis/objkt/types';
+import { AttributeInfo } from '../interfaces/attribute.interface';
+import { CollectiblesRootState } from '../store/collectibles/collectibles-state';
+import {
+  CollectibleDetailsInterface,
+  CollectibleInterface
+} from '../token/interfaces/collectible-interfaces.interface';
+import { getTokenSlug } from '../token/utils/token.utils';
+
+const attributesInfoInitialState: AttributeInfo[] = [
+  {
+    attributeId: 0,
+    tokens: 0
+  }
+];
+
+export const getAttributesWithRarity = (
+  attributesInfo: AttributeInfo[],
+  collectible: CollectibleInterface
+): CollectibleAttributes[] => {
+  const isExistGallery = isNonEmptyArray(collectible.galleries);
+  const collectibleCalleryCount = isExistGallery
+    ? collectible.galleries[0].gallery.items
+    : collectible.collection.items;
+
+  return collectible.attributes.map(({ attribute }) => {
+    const attributeTokenCount = attributesInfo.find(el => el.attributeId === attribute.id)?.tokens ?? 0;
+    const rarity = Number(((attributeTokenCount / collectibleCalleryCount) * 100).toFixed(2));
+
+    return {
+      attribute: {
+        id: attribute.id,
+        name: attribute.name,
+        value: attribute.value,
+        rarity
+      }
+    };
+  });
+};
+
+export const getAttributesInfo$ = (ids: number[], isGallery: boolean): Observable<AttributeInfo[]> => {
+  if (isGallery) {
+    return fetchGalleryAttributeCount$(ids).pipe(
+      map(result =>
+        result.map(({ attribute_id, tokens }) => ({
+          attributeId: attribute_id,
+          tokens
+        }))
+      ),
+      catchError(() => of(attributesInfoInitialState))
+    );
+  }
+
+  return fetchFA2AttributeCount$(ids).pipe(
+    map(result =>
+      result.map(({ attribute_id, tokens }) => ({
+        attributeId: attribute_id,
+        tokens
+      }))
+    ),
+    catchError(() => of(attributesInfoInitialState))
+  );
+};
+
+export const isAdultCollectible = (attributes?: CollectibleAttributes[], tags?: CollectibleTag[]) => {
+  let includesAdultAttributes = false;
+  let includesAdultTags = false;
+
+  if (isNonEmptyArray(attributes)) {
+    includesAdultAttributes = attributes.some(({ attribute }) => attribute.name === ADULT_ATTRIBUTE_NAME);
+  }
+
+  if (isNonEmptyArray(tags)) {
+    includesAdultTags = tags.some(({ tag }) => ADULT_CONTENT_TAGS.includes(tag.name));
+  }
+
+  return includesAdultAttributes || includesAdultTags;
+};
```

Comment:
```suggestion
export const isAdultCollectible = (attributes?: CollectibleAttributes[], tags?: CollectibleTag[]) => {
  let includesAdultAttributes = false;
  let includesAdultTags = false;

  if (isNonEmptyArray(attributes)) {
    includesAdultAttributes = attributes.some(({ attribute }) => attribute.name === ADULT_ATTRIBUTE_NAME);
  }

  if (isNonEmptyArray(tags)) {
    includesAdultTags = tags.some(({ tag }) => ADULT_CONTENT_TAGS.includes(tag.name));
  }

  return includesAdultAttributes || includesAdultTags;
};
```
```suggestion
export const isAdultCollectible = (attributes?: CollectibleAttributes[], tags?: CollectibleTag[]) => {
  if (attributes.some(({ attribute }) => attribute.name === ADULT_ATTRIBUTE_NAME)) {
    return true;
  }

  if (tags.some(({ tag }) => ADULT_CONTENT_TAGS.includes(tag.name))) {
    return true;
  }

  return false;
};
```

Reply:
We need to check on undefined attributes and tags. And I don't like the array method inside the condition